src/csharp/Grpc.Tools/implementation_notes.md

```diff
@@ -0,0 +1,219 @@
+
+The custom targets hook into various places in a normal MSBuild build by specifying
+before/after targets at the relevant places. See
+[msbuild-targets](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-targets)
+for predefined targets.
+
+* Before `PrepareForBuild`
+    * we add `Protobuf_SanityCheck` that checks this is a supported project type, e.g. a C# project.
+* Before `BeforeCompile`
+    * we add all the targets that compile the `.proto` files and generate the expected `.cs` files. These files are added to those that get compiled by the C# compiler.
+* After `CoreClean`
+    * we add `Protobuf_Clean` that cleans the files generated by the protobuf compiler.
+
+## Custom tasks
+
+There are a few custom tasks needed by the targets. These are implemented in C# in the Grpc.Tools project
+and packaged in the file `Protobuf.MSBuild.dll` in the NuGet package.
+
+* ProtoToolsPlatform
+    * Works out the operating system and CPU architecture
+* ProtoCompilerOutputs
+    * Tries to work out the names and paths of the files that would be generated by the protobuf compiler and returns these as a list of _items_
+    * Also returns list of _items_ that is the same as the _Protobuf_ items passed in with the output directory metadata updated
+* ProtoReadDependencies
+    * Read generated files from previously written dependencies file and return as items.
+* ProtoCompile
+    * Runs the protobuf compiler for a proto file. The executable to run is specified by the property `Protobuf_ProtocFullPath`
+    * To do this it:
+        * first writes out a response file containing the parameters for protobuf compiler 
+        * runs the executable, which generates the `.cs` files and a `.protodep` dependencies file
+        * reads the dependencies file to find the files that were generated and these are returned as a list of _items_ that can then be used in the MSBuild targets
+
+## Build steps
+
+The names of these items and properties are correct at the time of writing this document.
+
+High level builds steps:
+
+* Prepare list of `.proto` files to compile
+    * Makes sure all needed metadata is set for the `<Protobuf>` item, defaulting some values
+    * Removes `<Protobuf>` items that no longer exist or are marked as don't compile
+* Handling incremental builds
+    * Work out files that need to be created or have changed
+* Compile the `.proto` files
+* Add generated files to the list of files for the C# compiler
+
+### Prepare the list of .proto files to compile
+
+At various stages of the build copies of the original `<Protobuf>` items are created
+and/or updated to set metadata and to prune out unwanted items.
+
+Firstly, the build makes sure `ProtoRoot` metadata is set for all `Protobuf` items.
+A new list of items - `Protobuf_Rooted` - is created from the `Protobuf` items with `ProtoRoot` metadata set
+
+* Either already explicitly set in the item
+* Or "." if in project directory
+* Or relative path to project directory
```

Comment:
rephrase: If the .proto file outside of project's directory set `ProtoRoot="...."`. Might wanna rephrase "relative path to project directory" for clarity.